src/object.d

```diff
@@ -3287,135 +3302,57 @@ int __cmp(T1, T2)(T1[] s1, T2[] s2)
-        }
 
-        auto len = s1.length;
-        if (s2.length < len)
-            len = s2.length;
-
-        auto fpArray1 = () { return cast(F[])s1; }();
-        auto fpArray2 = () { return cast(F[])s2; }();
-
-        foreach (const u; 0 .. len)
-        {
-            if (int c = compare(fpArray1[u], fpArray2[u]))
-                return c;
-        }
-        return s1.length < s2.length ? -1 : (s1.length > s2.length);
+        auto fpArray1 = () @trusted { return cast(F[])s1; }();
+        auto fpArray2 = () @trusted { return cast(F[])s2; }();
+        return __cmp(fpArray1, fpArray2);
     }
-    // char types = > dstrcmp
-    else static if ((is(U1 == ubyte) && is(U2 == ubyte))
-        || (is(U1 == void) && is(U2 == void))
-        || (is(U1 == bool) && is(U2 == bool))
-        || (is(U1 == char) && is(U2 == char)))
+    // All unsigned byte-wide types = > dstrcmp
+    else static if (is(U1 == ubyte) || is(U1 == void) || is(U1 == bool)
+        || is(U1 == char))
     {
-        if (!__ctfe)
-        {
-            import core.internal.string : dstrcmp;
-            return () @trusted { return dstrcmp(cast(char[])s1, cast(char[])s2); }();
-        }
-        else
-        {
-            // pretty ugly...
-            auto len = s1.length;
-
-            if (s2.length < len)
-                len = s2.length;
-
-            foreach (const u; 0 .. len)
-            {
-                auto e1 = at(s1, u);
-                auto e2 = at(s2, u);
-
-                if (e1 < e2)
-                    return -1;
-
-                if (e1 > e2)
-                    return 1;
-            }
-
-            return s1.length < s2.length ? -1 : (s1.length > s2.length);
-        }
+        import core.internal.string : dstrcmp;
+        return (() @trusted => dstrcmp(cast(char[])s1, cast(char[])s2))();
     }
-    // integral, struct, nested arrays
+    // Everything else
     else
     {
-        auto len = s1.length;
-
-        if (s2.length < len)
-            len = s2.length;
+        immutable len = s1.length <= s2.length ? s1.length : s2.length;
 
         foreach (const u; 0 .. len)
         {
-            auto e1 = at(s1, u);
-            auto e2 = at(s2, u);
-
             // structs
-            static if (__traits(compiles, e1.opCmp(e2)))
+            static if (__traits(compiles, at(s1, u).opCmp(at(s2, u))))
             {
-                auto c = e1.opCmp(e2);
+                auto c = at(s1, u).opCmp(at(s2, u));
                 if (c != 0)
-                    return c < 0 ? -1 : 1;
+                    return c;
             }
-            else static if (__traits(compiles, __cmp(e1, e2)))
+            else static if (__traits(compiles, __cmp(at(s1, u), at(s2, u))))
             {
-                auto c = __cmp(e1, e2);
+                auto c = __cmp(at(s1, u), at(s2, u));
                 if (c != 0)
-                    return c < 0 ? -1 : 1;
+                    return c;
             }
             else
```

Comment:
Hmm, doesn't look like it. Curiously, for a struct without `opCmp`, `S(1) < S(2)` doesn't compile, but `[S(1)] < [S(2)]` does. If S has a field with an `opCmp`, it doesn't get called. So I guess that's not something we need to worry about right now.

Though, there's still the question of structs without `opCmp`. I think the compiler just does a `memcmp` right now, but someone would need to check. Ideally we'd want to avoid doing that `memcmp` twice.